docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,151 @@
+
+```golang
+// GCIS stands for Global Candidate Index Set
+sort(GCIS, GCIS.Score, Descend_Order)
+recommend_set = GCIS[:N]
+remaining_set = GCIS[N:]
+min_cost = EvaluateWorkload(workload, recommend_set)
+
+for {
+    v_recommend_set, v_remaining_set = SwapElements(recommend_set, remaining_set, swap_size)
+    variant_cost = EvaluateWorkload(workload, v_recommend_set)
+    if variant_cost < min_cost {
+        min_cost = variant_cost
+        recommend_set = v_recommend_set
+        remaining_set = v_remaining_set
+    }
+    ......
+    // loop until specific rounds or timeout, break
+}
+```
+
+Note that executing `Swap and Re-evaluate` algorithm is necessary as the `reduced_cost` sometimes is a joint effect of several indexes and it's hard to tell each index's independent contribution to the final `reduced_cost`. For example, assume there is an extremely slow query in input workload and the desired indexes for this query is `a` and `b`. However, the number of allowed recommended indexes for the whole workload is limited and for some reason, `a` ranks top `n` in the final score list while `b` is not. But there are chances that without `b`, `a` can no more optimize that extremely slow query. 
+
+----------------------------------------------
+### A quick exmaple for single-column index recommendation
+
+**Workload**:
+
+```
+- Query1: SELECT COUNT(*) FROM t1 where a = 1 and b = 5
+- Query2: SELECT a FROM t1 ORDER BY c limit 10 
+- Query3: DELETE FROM t1 WHERE a = 1 AND d = 1
+```
+where index on column *c* has already existed. 
+
+- **Step1**: recommend indexes for single query. For each query, compile it and get the `original_cost` output by SQL optimizer. Then add all possible hypothetic indexes *a, b, d* to the schema, compile the same query again. After the compiling, `LSIs` can be extracted from output execution plan as well as the `virtual_cost`. Denote `reduced_cost` as *original_cost - virtual_cost*. Assume this step outputs:
+
+```
+Query 1--LSIs: {a, b}, reduced_cost: 4
+Query 2--LSIs: {c},    reduced_cost: 4
+Query 3--LSIs: {a, d}, reduced_cost: 3
+```
+- **Step2**: get the score list. Simply add up `LSIs`' `reduced_cost` of each query respectively and then sort the list descending by score. Assume this step outputs:
+
+```
+a: 4+3=7
+b: 4
+c: 4
+d: 3
+```
+
+- **Step3**: evaluate recommend index set. Assume the number of allowed indexes recommended by index advisor is *2* and the swap size is *1* 
+  - The initial `recommend index set` is `{a, b}`, the `remaining index set` is `{c,d}` 
+  - Add recommend index set `{a, b}` to the schema and at the same time mute existed index `c`
+  - Compile all queries in the workload and add up each query's cost. Assume the sum is *20* and then assign it variable `min_cost`
+  - Randomly exchange *1* element between the recommend set and remaining set. Assume *a* and *c* is chosen respectively, now the recommend set is `{c, b}` and the remaining set is `{a, d}`. 
```

Comment:
```suggestion
  - Randomly exchange *1* element between the recommended set and remaining set. Assume *a* and *c* are chosen respectively, now the recommended set is `{c, b}` and the remaining set is `{a, d}`. 
```

Reply:
`recommended` is a verb, so I'm afraid it's not appropriate for all `recommend (index) set` in this PR.